h2o-core/src/main/java/hex/grid/GridSearch.java

```diff
@@ -205,26 +208,40 @@ public void onBuildFailure(final ParallelModelBuilder.ModelBuildFailure modelBui
       } finally {
         parallelSearchGridLock.unlock();
       }
-      attemptBuildNextModel(parallelModelBuilder, null);
+    }
+
+    @Override
+    public void afterBuildProcessing(ParallelModelBuilder parallelModelBuilder, Model finishedModel) {
+      attemptBuildNextModel(parallelModelBuilder, finishedModel);
     }
 
     private void attemptBuildNextModel(final ParallelModelBuilder parallelModelBuilder, final Model previousModel) {
-      // Attempt to train next model
-      final MP nextModelParams = getNextModelParams(hyperspaceIterator, previousModel, grid);
-      if (nextModelParams != null
-              && isThereEnoughTime()
-              && !_job.stop_requested()
-              && !_hyperSpaceWalker.stopEarly(previousModel, grid.getScoringInfos())) {
 
-        // FIXME: ModelFeeder's <MP> type parameter is hiding GridSearch's one. Ideally would have moved ModelFeeder outside but it is too coupled with GridSearch class.
-        reconcileMaxRuntime(grid._key, hyperspaceIterator, nextModelParams);
+      synchronized (hyperspaceIterator) {
+        if (hyperspaceIterator.hasNext(previousModel)
+                && isThereEnoughTime()
+                && !hasReachedMaxModels(parallelModelBuilder)
+                && !_job.stop_requested()
+                && !_hyperSpaceWalker.stopEarly(previousModel, grid.getScoringInfos())) {
 
-        parallelModelBuilder.run(Collections.singletonList(ModelBuilder.make(nextModelParams)));
-      } else {
-        parallelModelBuilder.noMoreModels();
+          final MP nextModelParams = getNextModelParams(hyperspaceIterator, previousModel, grid);
+
+          // FIXME: ModelFeeder's <MP> type parameter is hiding GridSearch's one. Ideally would have moved ModelFeeder outside but it is too coupled with GridSearch class.
```

Comment:
It's not hiding anything. Those are equal.

`class ModelFeeder<MP extends Model.Parameters, D extends ModelFeeder>` only defines it in the very same manner as the `GridSearch` class. If you want it to automatically change to whichever boundary is set to `MP` defined in `GridSearch`, you can just delete the `MP` generic parameter in `ModelFeeder` definition:

 `private class ModelFeeder<D extends ModelFeeder> `

Not sure if it helps anything.

Reply:
@Pscheidl please see my answer above. Having same name probably a convenient thing here for readability but it makes it impossible to assign those 2 types to each other.

Reply:
whats the point of the FIXME comment then? do you plan on fixing it, if yes, please add JIRA id to the comment, otherwise it will end up just like the other todos in the codebase

Reply:
@honzasterba I'm not going to fix this. The Jira like this would get forgotten I'm pretty sure. I just think that those todos and fixme things have a higher chance to be noticed by developers than forgotten Jiras. This is not a best practice for sure but our reality. I will just remove the comment.

Reply:
will just rename ModelFeeder's type parameter

Reply:
some more reading on this subject: https://wiki.c2.com/?TodoCommentsConsideredHarmful

if even a jira will not make you fix it than just forget about it

Reply:
@honzasterba thanks, I know this. I just think that we always should consider situation at hand for a final decision. Anyway, fixed